Do partitioning and sorting in Crunch

tomwhite opened the pull request:
This change gives a marked performance increase (e.g. from 28 to 14 minutes on ALL.chr22.phase1_release_v3.20101123.snps_indels_svs.genotypes.vcf).

It relies on an extension to Crunch that produces one file per partition of a PTable<String, T>, where the key is the file name. This already exists in Crunch for Avro files, but not for Parquet, so I've added that in https://issues.apache.org/jira/browse/CRUNCH-562. Until this is released in Crunch I've included the files in this PR too.

This change also fixes #11 by using Hive commands to create the table and add partitions.

tomwhite commented:
Made a change to flatten and expand calls after the shuffle, and got another good speedup (to around 6 minutes now).

tomwhite commented:
Flattening after the shuffle doesn't work, since each partition need to be sorted by sample ID. I've changed this back and added some tests to ensure the sort works. I've also included a change to support multiple VCFs.

laserson commented on README.md:
````diff
@@ -63,20 +47,20 @@ hadoop jar target/quince-0.0.1-SNAPSHOT-job.jar \
 
 ```bash
-hive -e 'show create table datasets.variants_flat_locuspart'
+hive -f sql/create-variants.sql --hiveconf namenode='bottou01.sjc.cloudera.com'
````
Possible for the schema to be inferred from the Parquet data set instead of hardcoding the schema?  That would be more future-proof as we add possible data types or support flat and nested data.  Using something like Impala's `CREATE TABLE foo LIKE PARQUET ...`.  (Why isn't that DDL stmt available in Hive?)

tomwhite replied:
We might be able to do that. There's a slight wrinkle in that you have to find the name of the file since it could be in any chr=?/pos=? directory. I'd like to tackle this later if that's OK.

laserson replied:
Sure. Just create an issue as a reminder

tomwhite replied:
https://github.com/cloudera/quince/issues/22

laserson commented on README.md:
````diff
@@ -139,16 +123,26 @@ For Impala, update the partition information with:
+```bash
+hive -e 'drop table datasets.variants_flat_locuspart'
+hadoop fs -rm -r datasets/variants_flat_locuspart
+```
````
What about dropping just a single partition/sample group?

tomwhite replied:
We could add that as well I think. I tried running the following but hit some problems. Opened https://github.com/cloudera/quince/issues/21

```
hive -e "truncate table datasets.variants_flat_locuspart partition (sample_group='sample1')"
hadoop fs -rm -r datasets/variants_flat_locuspart/chr=*/pos=*/sample_group=sample1
```

laserson commented on src/main/java/com/cloudera/science/quince/CrunchUtils.java:
```diff
@@ -0,0 +1,126 @@
+    public void process(FlatVariantCall input,
+        Emitter<Pair<String, Pair<String, FlatVariantCall>>> emitter) {
+      String partitionKey = extractPartitionKey(input, segmentSize, sampleGroup);
+      emitter.emit(Pair.of(partitionKey, Pair.of(input.getCallSetId().toString(), input)));
```
In the interest of maximally homogenizing the Parquet pages, does it makes sense to do a secondary sort on `(sample_id_id, chr, pos)` instead of just `(sample_id,)`?

tomwhite replied:
Done

laserson commented on src/main/java/com/cloudera/science/quince/CrunchUtils.java:
```diff
@@ -0,0 +1,126 @@
+        new FlattenVariantFn(), Avros.specifics(FlatVariantCall.class));
+    // group by partition key (table key), then prepare for sorting by sample ID (first
+    // element in value pair)
+    PTable<String, Pair<String, FlatVariantCall>> partitionedRecords =
```
maybe "`keyedRecords`" is more clear?

tomwhite replied:
Done

laserson commented:
I needed to change `com.github.samtools:htsjdk` to version 1.138 for this to compile on my machine.

laserson commented:
Why can't the flattening post-shuffle work?  It is horrendous to simply do a groupby operation on the partition keys and then sort each of the resulting partition's iterables on the post-shuffle side?

laserson commented on src/main/java/com/cloudera/science/quince/CrunchUtils.java:
```diff
@@ -0,0 +1,126 @@
+   * Turns a (partition key, list[sample id, variant call]) pair into a series of
+   * (partition key, flat variant call) pairs, expanding the calls (samples) in the variant.
+   */
+  private static final class ExtractFlatVariantsFn extends
```
Should we delete this if it's not being used?

laserson commented on src/main/java/com/cloudera/science/quince/VariantContextToVariantFn.java:
```diff
@@ -42,47 +47,55 @@
-      byte[] variantHeaders = Base64.decodeBase64(getConfiguration().get(VARIANT_HEADER));
-      VcfBlockIterator iterator = new VcfBlockIterator(
-          new ByteArrayInputStream(variantHeaders), new FullVcfCodec());
+    converter = new VariantContext2VariantConverter();
```
This initialization should occur per mapper.  Would it not be possible to read the VCF header here just for the file that it being processed?  This eliminates the need to read all the headers in advance, and serialize them into the Configuration.  Also seems cleaner to me.  You think this would add a lot of overhead?

tomwhite replied:
I'm not sure if this is worth the effort, as the headers are pretty small. Also, detecting which file a mapper is working on is non-trivial.

laserson commented on src/main/java/com/cloudera/science/quince/VariantContextToVariantFn.java:
```diff
@@ -42,47 +47,55 @@
+  private static final String VARIANT_HEADERS = "variantHeaders";
+  private static final String VARIANT_SET_ID = "quinceVariantSetId";
+
+  public static void configureHeaders(Configuration conf, Path[] vcfs, String sampleGroup)
```
Possible to eliminate this static method, per comment below?

laserson commented on src/main/java/org/apache/crunch/io/parquet/AvroParquetPathPerKeyOutputFormat.java:
```diff
@@ -0,0 +1,95 @@
+      String key = record.datum().key().toString();
+      if (!key.equals(currentKey)) {
+        if (currentWriter != null) {
+          currentWriter.close(null); // TaskAttemptContext not used for close
```
What's the significance of this comment?  (I guess I'm asking for a meta-comment :) )

tomwhite replied:
This code is going in to Crunch, so it needs addressing there (basically, passing in a null TaskAttemptContext is OK since it is not called in the close method).

laserson commented on src/main/java/org/apache/crunch/io/parquet/AvroParquetPathPerKeyOutputFormat.java:
```diff
@@ -0,0 +1,95 @@
+ *
+ * This {@code OutputFormat} only keeps one {@code RecordWriter} open at a time, so it's a very good idea to write
+ * out all of the records for the same key at the same time within each partition so as not to be frequently opening
+ * and closing files.
```
Assuming the keys weren't sorted, if you go back to an earlier key, it will simply recreate a new file, yes?  So if my partition had two unique keys, and they were alternating from record to record, then I would get a new file for each record?

tomwhite replied:
That's a limitation of this class, but one that we are not encountering since we do sort by key.

laserson commented on src/test/java/com/cloudera/science/quince/VariantContextToVariantFnTest.java:
```diff
@@ -0,0 +1,80 @@
+    VariantContextWritable vcw = new VariantContextWritable();
+    vcw.set(vc);
+
+    class CapturingEmitter implements Emitter<Variant> {
```
Should this be made into a general utility?

tomwhite replied:
Maybe for testing? On the other hand, it may be better just to use Mockito.

laserson commented:
Maybe changes the default `sampleGroup` in the CLI to `"default"`?

laserson commented:
Generally looks good to me!

tomwhite commented:
I've addressed most of the issues you raised @laserson. Regarding doing the flattening post shuffle, I've a branch to try that out, so that can go in later.
